### PR TITLE
fix(workflow-edges): enforce edge/block validation server-side, not just client-side

### PR DESCRIPTION
**Pull request by waleedlatif1**
## Summary
- Dragging a connection that creates a cycle correctly refused to render client-side, but the cyclic edge was still queued for realtime persistence and written to `workflow_edges`, so it reappeared after refresh
- Root cause: cycle detection (`wouldCreateCycle`) — and, on audit, several other edge/block validation rules (scope-boundary crossing, annotation-only-block edges, duplicate edges, trigger-block targets, reserved/duplicate block names) — only ever ran in the client Zustand store, never in `apps/realtime`'s DB write path, which is the actual source of truth on reload
- Moved these into `@sim/workflow-types` (importable by both `apps/sim` and `apps/realtime`) so the client store, the collaborative queueing layer, and realtime's persistence layer all share one implementation instead of independently (and silently) drifting
- Wired the shared checks into `apps/realtime/src/database/operations.ts`'s edge-add (single + batch) and block-rename handlers as the authoritative gate
- Client-side call sites, error messages, and rule ordering are unchanged — this is a server-side enforcement gap fix, not a UX change

## Type of Change
- [x] Bug fix

## Testing
- `apps/sim/stores/workflows/workflow/store.test.ts`, `edge-validation.test.ts` (new), `utils.test.ts` — 139+ tests covering cycle rejection (single + same-batch), duplicate edges, scope-boundary edges, annotation-only edges, trigger-target edges, and reserved/duplicate block names
- Full executor test suite (1029 tests) passes, confirming the shared `normalizeName`/`RESERVED_BLOCK_NAMES` delegation is behavior-identical
- `tsc --noEmit` clean on `apps/sim` and `apps/realtime`; `bun run scripts/check-monorepo-boundaries.ts` and `check-realtime-prune-graph.ts` both pass
- Tested manually in the workflow editor: dragging Table → Agent to close a cycle no longer persists after refresh

## Checklist
- [x] Code follows project style guidelines
- [x] Self-reviewed my changes
- [x] Tests added/updated and passing
- [x] No new warnings introduced
- [x] I confirm that I have read and agree to the terms outlined in the [Contributor License Agreement (CLA)](./CONTRIBUTING.md#contributor-license-agreement-cla)

### CODE REVIEW

File: apps/realtime/src/database/operations.ts [hunk 1]
```diff
@@ -1331,11 +1374,68 @@ async function handleEdgeOperationTx(tx: any, workflowId: string, operation: str
...
+        isWorkflowAnnotationOnlyBlockType(blocksById[payload.source].type) ||
+        isWorkflowAnnotationOnlyBlockType(blocksById[payload.target].type)
+      ) {
+        logger.info('Skipping edge add - edge references an annotation-only block')
+        break
+      }
+
+      if (isKnownWorkflowTriggerBlock(blocksById[payload.target])) {
+        logger.info('Skipping edge add - trigger blocks cannot be edge targets')
+        break
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts]: Fixed in 878b0277f — added `triggerMode: workflowBlocks.triggerMode` to the block select in both the single-edge ADD and batch ADD queries. Good catch, this was a real gap: `isKnownWorkflowTriggerBlock` checked `block.triggerMode` but the column was never fetched, so the server-side trigger-target check was silently a no-op for triggerMode-toggled blocks.

File: apps/realtime/src/database/operations.ts [hunk 2]
```diff
@@ -1615,13 +1717,87 @@ async function handleEdgesOperationTx(
...
+
+      const uniqueEdges = filterUniqueWorkflowEdges(
+        structurallyValidEdges.map((e) => ({
+          source: e.source as string,
+          target: e.target as string,
+          sourceHandle: (e.sourceHandle as string | null) ?? null,
+          targetHandle: (e.targetHandle as string | null) ?? null,
+          original: e,
+        })),
+        existingEdgeEndpoints
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts]: Fixed in 878b0277f — made `filterUniqueWorkflowEdges` incremental (accumulates accepted edges into the comparison set as it processes the batch), matching the pattern already used by `filterAcyclicEdges`. Two duplicate edges within the same `BATCH_ADD_EDGES` payload now only let the first one through. Also added a regression test (`should not add duplicate connections that appear twice within the same batch`) covering this via the shared client store path.

File: apps/realtime/src/database/operations.ts [hunk 3]
```diff
@@ -1615,13 +1717,87 @@ async function handleEdgesOperationTx(
...
+          isWorkflowAnnotationOnlyBlockType(targetBlock.type)
+        ) {
+          countDrop('annotation-only block')
+          return false
+        }
+        if (isKnownWorkflowTriggerBlock(targetBlock)) {
+          countDrop('trigger block target')
+          return false
+        }
+        const scopeDropReason = getWorkflowEdgeScopeDropReason({ source, target }, blocksById)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts]: This is a known, intentional limitation — documented in the TSDoc on `isKnownWorkflowTriggerBlock` in packages/workflow-types/src/workflow.ts. The full client-side check (`TriggerUtils.isTriggerBlock`) has three conditions: `category === 'triggers'` (from the block registry), `triggerMode === true`, or the legacy `starter` type. Only the latter two are portable server-side — the registry-category classification only exists in `apps/sim/blocks`, which `apps/realtime` is architecturally forbidden from importing (enforced by `scripts/check-monorepo-boundaries.ts` and `check-realtime-prune-graph.ts` in CI).

I considered two workarounds and rejected both as hackier than leaving this gap documented:
1. Persisting a redundant `isTrigger`/`category` flag on the block row at write time — this reintroduces exactly the class of bug this PR fixes (a second copy of the same fact that can drift, e.g. if a block's registry category changes after the row was written).
2. Hand-maintaining a duplicate list of registry-category-trigger block types in the shared package — same drift risk, and silently goes stale as new trigger blocks are added to the registry.

Closing this fully would require either relaxing the realtime/executor package boundary or making the block registry itself boundary-safe (pure metadata split out from React/Next-coupled code) — that's a real, larger architectural project, not something to bolt on here. Left a pointer comment in the code; happy to file a follow-up issue for it if that's useful.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/realtime/src/database/operations.ts [hunk 4]
```diff
@@ -1615,13 +1719,87 @@ async function handleEdgesOperationTx(
...
+        .from(workflowEdges)
+        .where(eq(workflowEdges.workflowId, workflowId))
+      const existingEdgeEndpoints = existingEdgesForCycleCheck.map(toEdgeHandles)
+
+      const uniqueEdges = filterUniqueWorkflowEdges(
+        structurallyValidEdges.map((e) => ({
+          source: e.source as string,
+          target: e.target as string,
+          sourceHandle: (e.sourceHandle as string | null) ?? null,
+          targetHandle: (e.targetHandle as string | null) ?? null,
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts]: Fixed in b959202db — the mismatch was between how the duplicate check normalized handles (`?? null`, so `''` and `null` compared as different) and how the insert path normalizes them (`|| null`, so `''` becomes `null`). Fixed at the single source of truth: `isDuplicateWorkflowEdge` in `@sim/workflow-types` now falsy-coalesces both sides before comparing, matching the persist-time normalization exactly. Added a regression test covering a `sourceHandle: ''` edge deduping against a same-source/target edge with no handle.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

**Comment by waleedlatif1** [comment]
Ran an independent /simplify pass (4 parallel review agents: reuse, simplification, efficiency, altitude) before treating this as done, on top of the already-clean Greptile 5/5 / Cursor Bugbot review:

- **Efficiency: clean, no changes.** Verified the sequential blocksById-then-edges query ordering in the realtime handlers is a deliberate cheap-work-first short-circuit (not a missed `Promise.all`), and the DFS cycle check / batch dedup are correctly bounded for this domain's actual scale (tens–hundreds of blocks/edges).
- **Simplification + altitude (real finding, fixed in c1ce0a073):** `handleEdgeOperationTx` (single ADD) and `handleEdgesOperationTx` (BATCH_ADD_EDGES) in `operations.ts` hand-inlined the identical 6-step validation pipeline and each re-fetched blocksById/existingEdgesForCycleCheck independently — two copies of one rule in the same file, which is exactly the client/server drift class this PR closes everywhere else. Extracted `filterEdgesForPersist` so both call one path. Net **-63 lines**. This also fixed a real bug it surfaced: `droppedCounts` was keyed by the scope-boundary check's free-text, block-id-bearing message, so it could never aggregate across edges — now keyed by a stable `'scope boundary'` reason.
- **Reuse (real finding, fixed in c1ce0a073):** `use-collaborative-workflow.ts`'s `collaborativeUpdateBlockName` still hand-rolled the empty/reserved/duplicate name-conflict checks this PR centralized as `getWorkflowBlockNameConflict` (already adopted by `store.ts`). Switched it over — which also fixed a latent check-order mismatch (this pre-check ran reserved-before-duplicate; `store.ts`'s real gate runs duplicate-before-reserved after this PR, so the two could disagree on which toast a name that was both reserved and duplicate would show).
- **Altitude, no code change, flagging for the record:** the trigger-target check on the realtime path (`isKnownWorkflowTriggerBlock`) only catches the legacy `triggerMode`/`starter` case, not modern category-based trigger blocks — that's a documented, deliberate architectural boundary (apps/realtime can't import apps/sim/blocks), not a bandaid. Worth a one-line caveat in the PR description that server-side enforcement closes the gap for legacy trigger targets specifically, not full parity with client-side trigger detection.

Re-verified after the fix: `type-check`, `lint`, monorepo-boundaries check, and all 112 apps/realtime tests + 91 apps/sim workflow-store/edge-validation tests pass.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/realtime/src/database/operations.ts [hunk 5]
```diff
@@ -24,14 +24,187 @@ import {
...
+  const existingEdgesForCycleCheck: PersistedEdgeRecord[] = await tx
+    .select({
+      sourceBlockId: workflowEdges.sourceBlockId,
+      targetBlockId: workflowEdges.targetBlockId,
+      sourceHandle: workflowEdges.sourceHandle,
+      targetHandle: workflowEdges.targetHandle,
+    })
+    .from(workflowEdges)
+    .where(eq(workflowEdges.workflowId, workflowId))
+  const existingEdgeEndpoints = existingEdgesForCycleCheck.map(toEdgeHandles)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts:195]: Verified this empirically before responding, not just by re-reading the code.

This specific race can't happen here: `persistWorkflowOperation` (the sole caller of `handleEdgeOperationTx`/`handleEdgesOperationTx`) opens `await db.transaction`, and the *first* statement in every transaction is `UPDATE workflow SET updatedAt = ... WHERE id = workflowId`. That's a row-level lock on the parent `workflow` row, held for the transaction's full duration — so two concurrent `persistWorkflowOperation` calls for the same `workflowId` cannot interleave; the second blocks on that very first UPDATE until the first transaction commits or rolls back. `filterEdgesForPersist`'s read-then-insert sequence runs entirely inside that already-serialized window.

I confirmed this isn't just theoretical: spun up a throwaway local Postgres and ran two concurrent transactions matching this exact shape (UPDATE the parent row → sleep 3s to simulate the read/validate window → insert → commit). The second transaction's UPDATE blocked for the full 3s and only returned the instant the first committed. So there's no gap for two writers to both read the same pre-insert edge set.

Rather than add a second, redundant advisory lock on top of an already-serializing UPDATE, I added a comment at that UPDATE (a7fdb6ade) documenting that it's this workflow's write-serialization point, not just a timestamp bump — so a future change can't silently remove that guarantee by making it conditional as a perceived no-op optimization. Appreciate the catch either way — it's a completely reasonable thing to flag from the diff alone, since the guard lives several dozen lines away in the calling function, not next to `filterEdgesForPersist` itself.

**Comment by waleedlatif1** [comment]
Two more things surfaced while tracing every `workflowEdges` insert site in `operations.ts` after Cursor's summary flagged `BATCH_ADD_BLOCKS` — one fixed (f75cd5e4c), one deliberately **not** fixed here:

**Fixed:** `BATCH_ADD_BLOCKS` (block paste/duplicate/import) inserted its edges with zero validation — confirmed by reading the handler, not just from the summary. Routed through `filterEdgesForPersist`, same as the other two edge-add paths.

**Found, not fixed — `WORKFLOW_OPERATIONS.REPLACE_STATE`:** this operation (full graph replace, sent via `enqueueReplaceWorkflowState` in `apps/sim/lib/workflows/operations/socket-operations.ts`) also bulk-inserts edges unvalidated. It's real and client-reachable, not dead code — the client does call `normalizeWorkflowState` before enqueuing, but that only runs `validateEdges` (missing-block/protected/annotation/scope), not the cycle or duplicate checks, and nothing stops a client from emitting the raw socket operation directly, bypassing that anyway (the entire premise of this PR). I'm not fixing it in this PR: this operation deletes and replaces the *entire* workflow atomically, and I don't have enough certainty about its callers' failure-mode expectations (undo/redo restore, version revert, diff-apply) to know whether silently dropping invalid edges — the pattern used everywhere else in this file — is correct here, versus rejecting the whole replace, versus something else. Guessing at that and shipping it would be worse than leaving it open. Flagging clearly so it gets tracked and fixed deliberately rather than silently patched.

**Noted, not touched — `insertAutoConnectEdge` (line ~271):** also unvalidated, but it's genuinely dead code — no callers anywhere in `apps/realtime` or `apps/sim`. Not a live gap, but worth deleting in a follow-up so it doesn't get wired up later without anyone noticing it lacks validation.

Re-verified after the `BATCH_ADD_BLOCKS` fix: `type-check`, `lint`, monorepo-boundaries, and all 112 apps/realtime tests pass.

**Comment by waleedlatif1** [comment]
@greptile

**Comment by waleedlatif1** [comment]
@cursor review

File: apps/realtime/src/database/operations.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts:195]: Acknowledged as a real finding, but this is a pre-existing race (the original code before this PR had the same read-check-insert pattern with no serialization or DB-level uniqueness constraint on the logical edge identity) — not something introduced by this PR's changes. Closing it properly needs either a DB-level unique constraint (schema migration) or an application-level lock (e.g. a Postgres advisory lock scoped per workflow), both of which are a different class of change than the client/server validation-drift bug this PR set out to fix. Deferring — will track as a follow-up rather than bundling a new concurrency-control pattern into this PR.

[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/realtime/src/database/operations.ts:195]: Correction to my earlier reply — I was wrong that this needed a new mechanism. Verified (including empirically against a local Postgres): `persistWorkflowOperation`'s leading `UPDATE workflow SET updatedAt ... WHERE id = workflowId` (which every operation, including edge adds, runs first inside the same transaction) already takes a row lock on `workflow` for the rest of the transaction. A second concurrent `persistWorkflowOperation` for the same `workflowId` blocks on that UPDATE until the first transaction commits or rolls back — so `filterEdgesForPersist`'s validate-then-insert sequence can't actually interleave across two writers on the same workflow. No code change needed; documented the invariant at the lock site (operations.ts:405-413) so a future 'optimize away this seemingly-redundant UPDATE' change can't silently reopen the race.

**Comment by waleedlatif1** [comment]
@greptile
